tests/unit/test_refine.py

```diff
@@ -280,160 +103,92 @@ def test_valid_coreset_no_kernel(self):
-        a toy example:
-
-        For ``X = [[0,0], [1,1], [2,2]]``, the 2-point coreset that minimises the MMD is
-         specified by the indices ``coreset_indices = [0, 2]``, i.e.
-         ``X_c =  [[0,0], [[2,2]]``.
-
-        We test, when given ``coreset_indices=[2,2]``, that ``refine()`` updates the
-        coreset indices to ``[0, 2]``.
-        """
-        original_array = jnp.asarray([[0, 0], [1, 1], [2, 2]])
-        best_indices = {0, 2}
-        test_indices = [2, 2]
-        coreset_indices = jnp.array(test_indices)
-        kernel = coreax.kernel.SquaredExponentialKernel()
-        coreset_obj = CoresetMock(weights_optimiser=None, kernel=kernel)
-        coreset_obj.coreset_indices = coreset_indices
-        coreset_obj.original_data = coreax.data.ArrayData.load(original_array)
-        coreset_obj.coreset = original_array[coreset_indices, :]
-        coreset_obj.kernel_matrix_row_sum_mean = None
-        refine_rand = coreax.refine.RefineRandom(self.random_key, p=1.0)
-
-        with patch.object(
-            coreax.kernel.Kernel,
-            "calculate_kernel_matrix_row_sum_mean",
-            wraps=kernel.calculate_kernel_matrix_row_sum_mean,
-        ) as mock_method:
-            refine_rand.refine(coreset=coreset_obj)
-
-        # Check the approximation method in the Kernel class is called exactly once
-        mock_method.assert_called_once()
+        with pytest.raises(AttributeError, match="object has no attribute 'compute'"):
+            refine_method.refine(coreset=coreset_obj)
 
-        self.assertSetEqual(set(coreset_obj.coreset_indices.tolist()), best_indices)
-
-    def test_invalid_coreset(self):
-        """
-        Test how RefineRandom handles an invalid coreset input.
-        """
+    def test_invalid_coreset_argument(self, refine_method: coreax.refine.Refine):
+        """Test behaviour of the ``refine`` method when passed an invalid coreset."""
         # Define an object to pass that is not a coreset, and does not have the
         # associated attributes required to refine
-        refine_random = coreax.refine.RefineRandom(self.random_key, p=0.5)
-        with self.assertRaises(AttributeError) as error_raised:
-            refine_random.refine(coreset=coreax.util.InvalidKernel(x=1.0))
+        with pytest.raises(
+            AttributeError, match="object has no attribute 'validate_fitted'"
+        ):
+            refine_method.refine(coreset=coreax.util.InvalidKernel(x=1.0))
+
+    @pytest.mark.parametrize(
+        "array, test_indices, best_indices",
+        [
+            (
+                jnp.asarray([[0, 0], [1, 1], [0, 0], [1, 1]]),
+                list(itertools.combinations(range(4), 2)),
+                [{0, 1}, {0, 3}, {1, 2}, {2, 3}],
+            ),
+            (jnp.asarray([[0, 0], [1, 1], [2, 2]]), [[2, 2]], [{0, 2}]),
+        ],
+    )
+    @pytest.mark.parametrize("use_cached_row_sum_mean", [False, True])
+    def test_refine(
```

Comment:
The parametrisation of this test is hard to parse - it needs more explanation attached, and might be better off split into two separate tests (as the tests they come from are testing subtly different things).

Reply:
Looking at this again, I totaly agree. This is confusing, especially as it wasn't clear that this addressed the `test_refine_ones`. I'll have a think about a nicer way to do this.